src/block_helpers.rs

```diff
@@ -88,39 +111,81 @@ pub fn read_weights_only_from_buf2<L:OptimizerTrait>(weights_len: usize, out_wei
                     fb: &feature_buffer::FeatureBuffer, 
+                    pb: &mut port_buffer::PortBuffer,                             
                     update: bool) -> f32 {
 
-    unsafe {
-        let block_loss_function: Box<dyn BlockTrait> = mem::transmute(& *block_loss_function.deref().deref());
-        let mut further_blocks_v: Vec<Box<dyn BlockTrait>> = vec![block_loss_function];
-        let further_blocks = &mut further_blocks_v[..];
-        let (prediction_probability, general_gradient) = block_run.forward_backward(further_blocks, 0.0, fb, update);
-        // black magic here: forget about further blocks that we got through transmute:
-        further_blocks_v.set_len(0);
+        pb.reset();
+        let (block_run, further_blocks) = bg.blocks_final.split_at_mut(1);
+        block_run[0].forward_backward(further_blocks, fb, pb, update);
+        let prediction_probability = pb.observations[0];
         return prediction_probability
-    }
 }
 
-/// This function is used only in tests to run a single block with given loss function
-pub fn spredict<'a>(block_run: &mut Box<dyn BlockTrait>, 
-                    block_loss_function: &mut Box<dyn BlockTrait>,
+pub fn spredict2<'a>(bg: &mut graph::BlockGraph,
                     fb: &feature_buffer::FeatureBuffer, 
+                    pb: &mut port_buffer::PortBuffer,                             
                     update: bool) -> f32 {
 
-    unsafe {
-        let block_loss_function: Box<dyn BlockTrait> = mem::transmute(& *block_loss_function.deref().deref());
-        let mut further_blocks_v: Vec<Box<dyn BlockTrait>> = vec![block_loss_function];
-        let further_blocks = & further_blocks_v[..];
-        let prediction_probability = block_run.forward(further_blocks, 0.0, fb);
-        // black magic here: forget about further blocks that we got through transmute:
-        further_blocks_v.set_len(0);
+        pb.reset();
+        let (block_run, further_blocks) = bg.blocks_final.split_at(1);
+        block_run[0].forward(further_blocks, fb, pb);
+        let prediction_probability = pb.observations[0];
         return prediction_probability
+}
+
+
+#[inline(always)]
+pub fn forward_backward(further_blocks: &mut [Box<dyn BlockTrait>], 
+                        fb: &feature_buffer::FeatureBuffer, 
+                        pb: &mut port_buffer::PortBuffer, 
+                        update:bool) {
+    match further_blocks.split_first_mut() {
+        Some((next_regressor, further_blocks)) => next_regressor.forward_backward(further_blocks, fb, pb, update),
+        None => {},
     }
 }
 
+#[inline(always)]
+pub fn forward(         further_blocks: &[Box<dyn BlockTrait>], 
+                        fb: &feature_buffer::FeatureBuffer, 
+                        pb: &mut port_buffer::PortBuffer) {
+    match further_blocks.split_first() {
+        Some((next_regressor, further_blocks)) => next_regressor.forward(further_blocks, fb, pb),
+        None => {},
+    }
+}
+
```

Comment:
Perhaps worth removing these extra spaces at the end of the file

Reply:
done